src/atdgen_codec_decode.ml

```diff
@@ -137,18 +129,16 @@ let obj_array f json = dict f json |> Js.Dict.entries
 let obj_list f json = obj_array f json |> Array.to_list
 
 let nullable decode json =
-  if (Obj.magic json : 'a Js.null) == Js.null then None else Some (decode json)
+  if Js.Json.test json Null then None else Some (decode json)
 
 (* Unlike Json_decode.field, this returns None if key is not found *)
 let fieldOptional key decode json =
-  if
-    Js.typeof json = "object"
-    && (not (Js.Array.isArray json))
-    && not ((Obj.magic json : 'a Js.null) == Js.null)
-  then
+  if Js.Json.test json Object then
     let dict = (Obj.magic (json : Js.Json.t) : Js.Json.t Js.Dict.t) in
     match Js.Dict.get dict key with
     | None -> None
+    (* treat fields with null values as missing fields (atdgen's default) *)
+    | Some value when (Js.Json.test value Null) -> None
```

Comment:
I assume that before this change, the runtime would crash when receiving `null` values when in atd the field was defined as `option`? 😮

Reply:
Yes, it would attempt to decode `null` and throw an invalid field value error as if `keep_nulls` is set. I thought it was intended, at first, to require declaring potentially null values as `nullable`, until I realised the `keep_nulls` setting was supposed to be not set by default.